src/sentry/api/bases/external_actor.py

```diff
@@ -0,0 +1,128 @@
+from sentry.models import ExternalActor, Organization, OrganizationMember, Team, User
+from sentry.types.integrations import ExternalProviders, get_provider_choices
+
+AVAILABLE_PROVIDERS = {
+    ExternalProviders.GITHUB,
+    ExternalProviders.GITLAB,
+    ExternalProviders.SLACK,
+}
+
+
+class ExternalActorSerializerBase(CamelSnakeModelSerializer):  # type: ignore
+    external_id = serializers.CharField()
+    external_name = serializers.CharField(required=True)
+    provider = serializers.ChoiceField(choices=get_provider_choices(AVAILABLE_PROVIDERS))
+
+    @property
+    def organization(self):
+        return self.context["organization"]
+
+    def get_actor_id(self, validated_data: MutableMapping[str, Any]) -> int:
+        target_option = validated_data.pop(self._actor_key, None)
+
+        if not target_option:
+            raise Exception("Invalid actor")
+
+        return target_option.actor_id
+
+    def get_provider_id(self, validated_data: MutableMapping[str, Any]) -> int:
+        provider_name_option = validated_data.pop("provider", None)
+        return validate_provider(
+            provider_name_option, available_providers=AVAILABLE_PROVIDERS
+        ).value
+
+    def create(self, validated_data: MutableMapping[str, Any]) -> ExternalActor:
+        actor_id = self.get_actor_id(validated_data)
+        provider = self.get_provider_id(validated_data)
+
+        return ExternalActor.objects.get_or_create(
+            **validated_data,
+            actor_id=actor_id,
+            provider=provider,
+            organization=self.organization,
+        )
+
+    def update(
+        self, instance: ExternalActor, validated_data: MutableMapping[str, Any]
+    ) -> ExternalActor:
+        # Discard the object ID passed by the API.
+        if "id" in validated_data:
+            validated_data.pop("id")
+
+        for key, value in validated_data.items():
+            setattr(self.instance, key, value)
+        try:
+            self.instance.save()
+            return self.instance
+        except IntegrityError:
+            raise serializers.ValidationError(
+                "There already exists an external association with this external_name and provider."
+            )
+
+
+class ExternalUserSerializer(ExternalActorSerializerBase):
+    _actor_key = "user_id"
+
+    user_id = serializers.IntegerField(required=True)
+
+    def validate_user_id(self, user_id: int) -> User:
+        """ Ensure that this user exists and that they belong to the organization. """
+
+        organization_members = OrganizationMember.objects.filter(
+            user_id=user_id, organization=self.organization
+        ).select_related("user")
+
+        if not organization_members:
+            raise serializers.ValidationError("This member does not exist.")
+
+        return organization_members[0].user
+
+    class Meta:
+        model = ExternalActor
+        fields = ["user_id", "external_name", "provider"]
```

Comment:
I thought the `fields` were supposed to directly relate to the model (so was thinking it would `actor_id`), does the `user_id` and `team_id` for the team serializer need to be there if we've already explicitly defined it on the serializer?